plover_controller/machine.py

```diff
@@ -353,42 +352,38 @@ def complete_hat(self, hat: str):
             self._unsequenced_buttons_and_hats.add(f"{hat}{HAT_VALUES[value]}")
         del self._pending_hat_values[hat]
 
-    def maybe_complete_ordered_chord(self):
+    def any_active_inputs(self):
+      return (
+          any(abs(v) > self._params["stroke_end_threshold"] for v in self._stick_states.values()) or
+          any(v > 0 for v in self._trigger_states.values()) or
+          self._currently_pressed_buttons or
+          self._currently_uncentered_hats
+      )
+
+    def process_stick_movements(self):
+        if self.any_active_inputs():
+            return
+        def process(start_idx,end_idx):
+            if start_idx-end_idx == 0:
+                for key in pending_movements[start_idx:len(pending_movements)]:
+                    self._unsequenced_buttons_and_hats.add(key)
+            else:
+                result = self._mappings.ordered_mappings.get(tuple( pending_movements[start_idx:end_idx] ))
+                if result is not None:
+                    self._pending_keys.update(result)
+                    process(end_idx, len(pending_movements))
+                else:
+                    process(start_idx, end_idx-1)
```

Comment:
I'm hesitant to start (unconditionally) processing A B C D as the chord A B C and the single key D without understanding the motivation for crossing the deadzone in a stick motion in the first place. Do you have an example of a full steno mapping using this feature or some other idea for this change?

Reply:
Soz, I really should have elaborated more in my original post. It also had a couple bugs, oops.

## Motivation

This would increase the conveniently-accessible input space on the sticks. Namely it would allow fast repeat-direction motions like `left(l) → T; left(l,l) → T*`, or opposite-direction motions like `left(l) -> T; left(l,r) -> T*`. Maybe if you have an 8-way D-pad or are really accurate you could manage more precise movements too.
I personally use repeated-direction as an extra way to add asterisks, more comfortable to me than clicking the stick.
Across-direction motions are more likely to cause conflicts because you might literally want to use the two letters that are across from each other. If you already comfortably achieve all the sounds you need without touching the deadzone though, across-direction motions would be no problem and open up a lot of options for briefs.

This would also allow mapping misstrokes. I found when I try to make semicircles larger than 90° at high speed I might skip over the deadzone briefly. For example `left(ul,u,ur,r,lr)` gives me issues and it helps if I add misstrokes where I skip `u`, `ur`, or `r`: `left(ul,ur,r,lr); left(ul,u,r,lr); left(ul,u,ur,lr)`.

Note I'm only a < 10 WPM steno newb having fun theorycrafting, so I don't know how much these things will actually have an impact on speed.

## New maybe unintuitive behaviors

Example: `left(l) -> A; left(ul) -> B; left(u) -> C; left(ur) -> D`

Maybe unintuitive behavior 1: As you said, if the user has a mapping `left(l,ul,u) -> Z`, then a single smooth motion [l,ul,u,ur] would result in ZD, when the user might have expected it to default to the separate mappings and result in ABCD.
This could be solved by explicitly stating both `left(l,ul,u) -> Z` and `left(l,ul,u,ur) -> ABCD`. But you might have to worry about `left(l,ul,u,ur,r)`, `left(l,ul,u,ur,r,lr)` etc. all the way around too. I understand this could make mapping files larger, for people who prefer the original behavior.

Maybe unintuitive behavior 2: two separate but arpeggiated motions like [l,ul,u][ur] would be interpreted as one smooth motion [l,ul,u,ur] if the user has a mapping for `left(l,ul,u,ur)`, even though the motion was input in two parts.
I'm fine with this but if you don't like this behavior, I could fix it by keeping track of 1) when the stick goes into the deadzone and 2) which chords can be made in a single contiguous motion. Then if a chord can be made in one contiguous motion, it should _not_ activate if the user touched the deadzone somewhere in the middle of it.

## Example Mapping

Here is a mapping I'm working on for myself (WIP) which demonstrates same-direction mappings to add an asterisk, and also misstrokes. I had to change the axes offsets and triggers for my Logitech F160 so you might have to change them depending on what controller you have. [my_mapping2.txt](https://github.com/user-attachments/files/16022270/my_mapping2.txt)

Below is a diagram of my left stick to help understand the mapping file:
<img src="https://github.com/tadeokondrak/plover_controller/assets/53129953/192da7a0-3fe7-4341-bd90-47e5ba5683ee" alt="left_stick_layout" width="500"/>

---

BTW I'm on an M1 Mac and plover_controller is working fine for me!

Reply:
My GitHub notifications are messed up so I'm only seeing this now. Thanks for the info! I'm ready to merge this once I look through and test the new code.

> BTW I'm on an M1 Mac and plover_controller is working fine for me!

This is really frustrating -- my Mac doesn't like my controller in any application for some reason, but I tried the same controller on a different Mac and it worked fine. One day I'll probably have to try factory resetting it.